File: 백준/Gold/14502. 연구소/연구소.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Map;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {

    static int n;
    static int m;
    static int[] dy = {-1, 1, 0, 0};
    static int[] dx = {0, 0, -1, 1};
    static int ans;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        int[][] arr = new int[n][m];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        buildWall(arr, 0, 0,  0);

        System.out.println(ans);




    }

    static void buildWall(int[][] arr, int y, int x, int cnt) {
        if (cnt == 3) {
            int[][] visited = new int[n][m];
            int[][] tempArr = new int[n][m];

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    tempArr[i][j] = arr[i][j];
                }
            }

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (tempArr[i][j] == 2 && visited[i][j] == 0) {
                        dfs(tempArr, visited, i, j);
                    }
                }
            }
            int result = 0;
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (tempArr[i][j] == 0) {
                        result++;
                    }
                }
            }
            ans = Math.max(ans, result);
            return;
        }
        int startI = y;
        int startJ = x;
        if (x == m) {
            startI++;
            startJ = 0;
        }
        for (int i = startI; i < n; i++) {
            int k = (i == startI) ? startJ : 0;
            for (int j = k; j < m; j++) {
                if (arr[i][j] == 0) {
                    arr[i][j] = 1;
                    buildWall(arr, i, j+1, cnt + 1);
                    arr[i][j] = 0;
                }
            }
        }
    }

    static void dfs(int[][] tempArr, int[][] visited, int y, int x) {
        visited[y][x] = 1;

        for (int i = 0; i < 4; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];

            if (ny < 0 || ny >= n || nx < 0 || nx >= m) continue;
            if (tempArr[ny][nx] == 0 && visited[ny][nx] == 0) {
                tempArr[ny][nx] = 2;
                dfs(tempArr, visited, ny, nx);
            }
        }
    }

}

File: 백준/Gold/14502. 연구소/연구소.py
import sys

n,m = map(int, sys.stdin.readline().rstrip().split())

arr = []
for _ in range(n):
    arr.append(list(map(int, sys.stdin.readline().rstrip().split())))
temp = [[0] * m for _ in range(n)] # 벽을 설치한 뒤의 맵 리스트

dx = [-1,0,1,0]
dy = [0,1,-0,-1]

ans = 0

# dfs 이용해 바이러스를 사방으로 퍼지게 하기
def virus(x,y):
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]
        if nx >= 0 and nx < n and ny >= 0 and ny < m:
            if temp[nx][ny] == 0:
                # 해당위치에 바이러스 배치 후 재귀
                temp[nx][ny] = 2
                virus(nx,ny)

# 현재 맵에서 안전 영역의 크기를 계산하기 함수
def get_score():
    score = 0

    for i in range(n):
        for j in range(m):
            if temp[i][j] == 0:
                score += 1
    return score

def dfs(count):
    global ans
    # 울타리가 3개 설치된 경우
    if count == 3:
        for i in range(n):
            for j in range(m):
                temp[i][j] = arr[i][j]
        # 각 바이러스의 위치에서 전파 진행
        for i in range(n):
            for j in range(m):
                if temp[i][j] == 2:
                    virus(i,j)
        # 안전 영역의 최댓값 계산
        ans = max(ans, get_score())
        return
    # 빈 공간에 울타리 설치
    for i in range(n):
        for j in range(m):
            if arr[i][j] == 0:
                arr[i][j] = 1
                count += 1
                dfs(count)
                arr[i][j] = 0
                count -= 1



dfs(0)
print(ans)

